File: examples/IO/number_game.hs
import System.Random
import Text.Read (readMaybe)

-- Retrieves a valid guess
getGuess :: IO Int
getGuess =
  putStr "Type your guess: " *>
  (readMaybe <$> getLine) >>= \ num ->
  case num of
    Just num | num >= 1 || num <= 100 -> pure num
    Nothing -> putStrLn "Come again?" *> getGuess
    _ -> putStrLn "Number needs to be from 1 to 100" *> getGuess

numberGame :: Int -> Int -> IO ()
numberGame 0 _ = putStrLn ("You lose! The number was " ++ show num ++ ".")
numberGame guesses num =
  putStrLn ("You have " ++ show guesses ++ " guesses left.") *>
  getGuess >>= \ guess ->
  case compare guess num of
    EQ -> putStrLn "Correct. You win!\n"
    LT -> putStrLn "Too low. Try again!\n"  *> numberGame (guesses - 1) num
    GT -> putStrLn "Too high. Try again!\n" *> numberGame (guesses - 1) num

main :: IO ()
main =
  putStrLn "I'm thinking of a number from 1 to 100. You have 7 guesses." *>
  (randomRIO (1, 100) >>= numberGame 7) *>
  putStrLn "Do you want to play again? (y/n)" *>
  getLine >>= \line -> case line of
    "y" -> putStr "\n" *> main
    "n" -> putStrLn "Goodbye."
    _ -> putStrLn "I'll take that as a yes!\n" *> main

File: examples/IO/number_game.py
#/usr/bin/env python3
import random

def getGuess():
    try:
        print("Type your guess: ")
        guess = int(input())
        if guess > 0 and guess <= 100:
            return guess
        else:
            print("Number needs to be from 1 to 100")
            return getGuess()
    except:
        print("Come again?")
        return getGuess()

def numberGame():
    guesses = 7
    number = random.randrange(1,100)
    print("I'm thinking of a number from 1 to 100. You have 6 guesses.")
    while guesses > 0:
        print("You have " + str(guesses) + " left.")
        guesses = guesses - 1

        guess = getGuess()
        if guess == number:
            print("Correct. You win!")
            return
        elif guess > number:
            print("Too high. Try again!")
        else:
            print("Too low. Try again!")
    print("You lose! The number was " + str(number) + ".")

while True:
    numberGame()
    print("Do you want to play again? (y/n)")
    answer = input()
    if answer == "n":
        break
    elif answer == "y":
        pass
    else:
        print("I'll take that as a yes!")
